src/main/java/org/tikv/common/util/ChannelFactory.java

```diff
@@ -26,80 +26,198 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.tikv.common.HostMapping;
 import org.tikv.common.pd.PDUtils;
 
 public class ChannelFactory implements AutoCloseable {
   private static final Logger logger = LoggerFactory.getLogger(ChannelFactory.class);
+  private static final String PUB_KEY_INFRA = "PKIX";
 
+  private int connRecycleTime;
   private final int maxFrameSize;
   private final int keepaliveTime;
   private final int keepaliveTimeout;
   private final int idleTimeout;
-  private final ConcurrentHashMap<String, ManagedChannel> connPool = new ConcurrentHashMap<>();
   private final CertContext certContext;
-  private final AtomicReference<SslContextBuilder> sslContextBuilder = new AtomicReference<>();
-  private static final String PUB_KEY_INFRA = "PKIX";
 
-  private abstract static class CertContext {
-    protected abstract boolean isModified();
+  private int epoch = 0;
+
+  @VisibleForTesting
+  public final ConcurrentHashMap<Pair<Epoch<SslContextBuilder>, String>, ManagedChannel> connPool =
+      new ConcurrentHashMap<>();
+
+  private final AtomicReference<Epoch<SslContextBuilder>> sslContextBuilder =
+      new AtomicReference<>();
+
+  private final ScheduledExecutorService recycler = Executors.newScheduledThreadPool(1);
+
+  private static class Epoch<T> {
+    private final T inner;
+    private final int epoch;
+
+    public Epoch(T inner, int epoch) {
+      this.inner = inner;
+      this.epoch = epoch;
+    }
+
+    public T get() {
+      return inner;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other == this) {
+        return true;
+      } else if (other instanceof Epoch) {
+        Epoch<?> o = (Epoch<?>) other;
+        return o.epoch == epoch;
+      } else {
+        return false;
+      }
+    }
+
+    @Override
+    public int hashCode() {
+      return epoch;
+    }
+  }
+
+  @VisibleForTesting
+  public static class CertWatcher implements AutoCloseable {
+    private static final Logger logger = LoggerFactory.getLogger(CertWatcher.class);
+    private final String path;
+    private final ArrayList<String> targets = new ArrayList<>();
+    private final BlockingDeque<Boolean> notify = new LinkedBlockingDeque<>();
+    private final ExecutorService executorService = Executors.newFixedThreadPool(2);
+
+    public CertWatcher(String path) {
+      if (path == null) {
+        throw new IllegalArgumentException("the base path of certs is missing");
+      }
+      this.path = path;
+      notify.push(true);
+      start();
+    }
+
+    private void watchLoop() {
+      try {
+        WatchService watchService = FileSystems.getDefault().newWatchService();
+        Path p = Paths.get(path);
+        p.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);
+
+        while (true) {
+          WatchKey key = watchService.take();
+          boolean changed = false;
+          OUTER:
+          for (WatchEvent<?> event : key.pollEvents()) {
```

Comment:
key.pollEvents() is non-blocking, which maybe waste CPU resource

Reply:
this method will not keep running, once the events were taken out of the queue, this block will exist and wait at `watchService.take()`

Reply:
Got it